pkg/networkserver/downlink.go

```diff
@@ -803,6 +809,227 @@ const gsScheduleWindow = 30 * time.Second
 // nsScheduleWindow is the time interval, which is sufficient for NS to ensure downlink is scheduled.
 var nsScheduleWindow = time.Second
 
+func (ns *NetworkServer) recordDownlink(dev *ttnpb.EndDevice, genDown *generatedDownlink, genState generateDownlinkState, down *scheduledDownlink) (nextDownlinkAt time.Time) {
+	if genState.ApplicationDownlink == nil || dev.MACState.LoRaWANVersion.Compare(ttnpb.MAC_V1_1) < 0 && genDown.FCnt > dev.Session.LastNFCntDown {
+		dev.Session.LastNFCntDown = genDown.FCnt
+	}
+	if genDown.NeedsAck {
+		dev.MACState.LastConfirmedDownlinkAt = timePtr(down.TransmitAt)
+	}
+	if genState.ApplicationDownlink != nil && genState.ApplicationDownlink.Confirmed {
+		dev.MACState.PendingApplicationDownlink = genState.ApplicationDownlink
+		dev.Session.LastConfFCntDown = genDown.FCnt
+	}
+	if dev.MACState.DeviceClass == ttnpb.CLASS_C {
+		var nextConfirmedAt time.Time
+		if dev.MACState.LastConfirmedDownlinkAt != nil {
+			nextConfirmedAt = dev.MACState.LastConfirmedDownlinkAt.Add(deviceClassCTimeout(dev, ns.defaultMACSettings))
+		}
+		if nextConfirmedAt.After(down.TransmitAt) {
+			nextDownlinkAt = nextConfirmedAt
+		} else {
+			nextDownlinkAt = down.TransmitAt.Add(dev.MACState.CurrentParameters.Rx1Delay.Duration())
+		}
+	}
+	dev.MACState.QueuedResponses = nil
+	dev.MACState.RxWindowsAvailable = false
+	dev.RecentDownlinks = appendRecentDownlink(dev.RecentDownlinks, down.Message, recentDownlinkCount)
+	return nextDownlinkAt
+}
+
+type downlinkAttemptResult struct {
+	applicationUpAppender func([]*ttnpb.ApplicationUp, bool) []*ttnpb.ApplicationUp
+
+	SetPaths       []string
+	NextDownlinkAt time.Time
+	Scheduled      bool
+	QueuedEvents   []events.Event
+}
+
+func (res downlinkAttemptResult) AppendApplicationUplinks(ups ...*ttnpb.ApplicationUp) []*ttnpb.ApplicationUp {
+	if res.applicationUpAppender == nil {
+		return ups
+	}
+	return res.applicationUpAppender(ups, res.Scheduled)
+}
+
+func (ns *NetworkServer) attemptClassADownlink(ctx context.Context, dev *ttnpb.EndDevice, phy band.Band, fp *frequencyplans.FrequencyPlan, maxUpLength uint16) downlinkAttemptResult {
+	var sets []string
```

Comment:
This one is not assigned to but used before L949, is that intentional?

Reply:
Yes.